sdk/core/core-auth/src/azureNamedKeyCredential.ts

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ * Represents a credential defined by a static API name and key.
+ */
+export interface NamedKeyCredential {
+  /**
+   * The value of the API key represented as a string
+   */
+  readonly key: string;
+  /**
+   * The value of the API name represented as a string.
+   */
+  readonly name: string;
+}
+
+/**
+ * A static name/key-based credential that supports updating
+ * the underlying name and key values.
+ */
+export class AzureNamedKeyCredential implements NamedKeyCredential {
+  private _key: string;
+  private _name: string;
+
+  /**
+   * The value of the key to be used in authentication.
+   */
+  public get key(): string {
+    return this._key;
+  }
+
+  /**
+   * The value of the name to be used in authentication.
+   */
+  public get name(): string {
+    return this._name;
+  }
+
+  /**
+   * Create an instance of an AzureNamedKeyCredential for use
+   * with a service client.
+   *
+   * @param name - The initial value of the name to use in authentication.
+   * @param key - The initial value of the key to use in authentication.
+   */
+  constructor(name: string, key: string) {
+    if (!name || !key) {
+      throw new TypeError("name and key must be non-empty strings");
+    }
+
+    this._name = name;
+    this._key = key;
+  }
+
+  /**
+   * Change the value of the key.
+   *
+   * Updates will take effect upon the next request after
+   * updating the key value.
+   *
+   * @param newName - The new name value to be used.
+   * @param newKey - The new key value to be used.
+   */
+  public update(newName: string, newKey: string): void {
```

Comment:
what if I just want to update one? Leaving the other empty seems like a not unreasonable contract compared to having to set it to its existing value

Reply:
In this case I went just based on the requirements for the linked issue:

> Allowing the name or key to be updated independently; both elements will be required when updating a credential.

I think it may make the API a bit confusing if we allow setting them separately.

I'd expect the `key` to change more frequently than the name (e.g. you've rotated your credentials for a given name.) I'd probably need to call `cred.update(undefined, newKey)` then, instead of `cred.update(cred.name, newKey)`. I can't think of a scenario where you'd change the name, but not the key.

I could obviously flip the position of name/key...but this would be opposite of what the other languages are doing, and mentally I think of the key being associated to the policy name, not the other way around.

I could expose `updateKey` and `updateName` methods, in addition to `update`, but that's 3 times the surface area now!

Reply:
I'm just a little confused, since I would expect updating the key to be way more common than the name, but perhaps I'm not understanding the name field's purpose very well.

I agree that it's better to not ship too much surface area at once, but are other languages also fine with this kinda clunky contract?

Reply:
@jsquire 
Pinging you since you created the issue :smile: What was the reason you didn't want to support updating name/key separately? Was it just to be concurrency safe?

Reply:
> What was the reason you didn't want to support updating name/key separately? Was it just to be concurrency safe?

Yes; the update/read is intended to be atomic to avoid race conditions with concurrent use.   For those languages with threads, allowing them to be set individually offers the chance for inconsistent data.

> but are other languages also fine with this kinda clunky contract?

This was the design that the architects approved, with the trade-offs made to favor safe concurrent use.  The goal was to ensure that we didn't provide overloads that callers may be tempted to use in a manner that could produce undesired results.  You may want to have a conversation with Brian since threading isn't a consideration here.

> perhaps I'm not understanding the name field's purpose very well.

The credential is intended for use when a service requires both the key name and key value to generate the value needed for authorization.  The name isn't meant to be informational, but part of the auth flow.   Not sure if that helps at all...  :smile:

> what if I just want to update one? Leaving the other empty seems like a not unreasonable contract compared to having to set it to its existing value

The intention is that updates will always provide both values;  essentially, you're fully reinitializing the state.   If the name is not provided, I would expect that the current value be updated to `undefined|null|whatever`.

> I'm just a little confused, since I would expect updating the key to be way more common than the name

I would agree.  However, it is a valid use case to generate a new access policy and need to replace the existing one, which requires the name and key.

Reply:
I'm once again relieved to not have to think about threading concerns in JS. :)